### PR TITLE
feature(analytics): add analytics local storage size limit and cleanup

### PR DESCRIPTION
**Pull request by phantumcode**
*Issue #, if available:*

*Description of changes:*
Add logic to check size of analytics local storage clean up storage to reduce size if it is over the limit

*Check points: (check or cross out if not relevant)*

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [ ] ~All integration tests pass~
- [x] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] ~Documentation update for the change if required~
- [x] PR title conforms to conventional commit style
- [ ] ~If breaking change, documentation/changelog update with migration instructions~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift [hunk 1]
```diff
@@ -108,3 +112,19 @@ final class SQLiteLocalStorageAdapter: SQLStorageProtocol {
 }
 
 extension SQLiteLocalStorageAdapter: DefaultLogger { }
+
+extension URL: DefaultLogger {
+    var attributes: [FileAttributeKey : Any]? {
```
Comment (ruisebas) [inline AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift:117]: I see how this is convenient, but it's also relying on a hard dependency to `FileManager` that can't be mocked during unit testing.

I suggest that we instead receive an instance of the existing `FileManagerBehaviour` in`SQLiteLocalStorageAdapter.init` that defaults to `FileManager.default`, and then expand that protocol to have the method you need.

Something like:
```swift
protocol FileManagerBehaviour {
    // ...
    func fileSize(for url: URL) -> Int
}

extension FileManager:  FileManagerBehaviour {
    // ...
    func fileSize(for url: URL) -> Int {
        // Grab the size from the attributes
    }
}

class SQLiteLocalStorageAdapter {
    // ...
    let fileManager: FileManagerBehaviour

    init(fileManager: FileManagerBehaviour = FileManager.default) { // + other arguments
        // ...
    }

    var bytesUsedOnDisk: Byte {
        return fileManager.fileSize(for: dbFilePath)
    }
}

```

Once you have a local variable, you can also completely remove all the other references to `FileManager.default`  in the  `SQLiteLocalStorageAdapter` class :)

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift [hunk 2]
```diff
@@ -14,10 +14,14 @@ final class SQLiteLocalStorageAdapter: SQLStorageProtocol {
     private var connection: Connection?
     private var dbFilePath: URL?
     
+    var diskByteUsed: Int {
```
Comment (ruisebas) [inline AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift:17]: nit: change `Bytes` to plural

**Review by ruisebas — CHANGES_REQUESTED** [review]

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift (hunk 2 above)
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift]: naming updated.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift (hunk 1 above)
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/SQLiteLocalStorageAdapter.swift]: Love the suggestion, code has been updated.

File: AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/AnalyticsEventSQLStorage.swift [hunk 3]
```diff
@@ -203,6 +193,19 @@ class AnalyticsEventSQLStorage: AnalyticsEventStorage {
         _ = try dbAdapter.executeQuery(moveStatement, [])
         _ = try dbAdapter.executeQuery(deleteStatement, [])
     }
+    
+    /// Check the disk usage limit of the local database.
+    /// If database is over the limit then delete all dirty events and oldest event 
+    /// - Parameter byteLimit: the size limit of the database
+    func checkDiskSize(byteLimit: Int) throws {
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AmplifyPlugins/Analytics/AWSPinpointAnalyticsPlugin/LocalStorage/AnalyticsEventSQLStorage.swift]: Thanks for pointing that out, I've updated to use the typealias `Byte`

**Review by ruisebas — APPROVED** [review]

**Pull request merged**